junit-jupiter-engine/src/test/kotlin/org/junit/jupiter/api/KotlinAssertionsTests.kt

```diff
@@ -211,6 +211,46 @@ class KotlinAssertionsTests {
         assertMessageStartsWith(error, assertionMessage)
     }
 
+    @Test
+    fun `assertNotNull with compiler smart cast`() {
+        val nullableString: String? = "string"
+
+        assertNotNull(nullableString)
+        assertFalse(nullableString.isEmpty()) // smart cast to non nullable object
+    }
+
+    @Test
+    fun `assertNull with compiler smart cast`() {
+        val nullableString: String? = null
+
+        assertNull(nullableString)
+        // even safe call is not allowed, because compiler knows that string is always null
+        // nullableString?.isEmpty()
+    }
+
+    @Test
+    fun `assertThrows with value initialization in lambda`() {
+        val value: String
+
+        assertThrows<AssertionError> {
+            value = "string"
+            Assertions.fail("message")
```

Comment:
Should this just be 
```suggestion
            fail("message")
```
considering this is Kotlin code and there's a `: Nothing` signature top level `fail`.

Reply:
In that case I agree with `UNKNOWN` for invocation kind, because not being able to call production code that declares `Nothing` as return type is a big restriction. The point of `assertThrows` is that it tests exceptions thrown from a method, and `Nothing` declares exactly that.

Reply:
In such case I'll set `InvocationKind.UNKNOWN` for `assertThrows` methods